src/main/java/org/folio/circulation/domain/LocationRepository.java

```diff
@@ -18,33 +16,72 @@
 import org.folio.circulation.support.Result;
 import org.folio.circulation.support.SingleRecordFetcher;
 
-import io.vertx.core.json.JsonObject;
-
 public class LocationRepository {
+
   private CollectionResourceClient locationsStorageClient;
+  private CollectionResourceClient institutionsStorageClient;
+  private CollectionResourceClient campusesStorageClient;
+  private CollectionResourceClient librariesStorageClient;
 
-  public LocationRepository(Clients clients) {
-    locationsStorageClient = clients.locationsStorage();
+  public LocationRepository(CollectionResourceClient locationsStorageClient,
+                            CollectionResourceClient institutionsStorageClient,
+                            CollectionResourceClient campusesStorageClient,
+                            CollectionResourceClient librariesStorageClient) {
+    this.locationsStorageClient = locationsStorageClient;
+    this.institutionsStorageClient = institutionsStorageClient;
+    this.campusesStorageClient = campusesStorageClient;
+    this.librariesStorageClient = librariesStorageClient;
   }
 
-  public CompletableFuture<Result<JsonObject>> getLocation(Item item) {
+  public static LocationRepository using(Clients clients) {
+    return new LocationRepository(
+      clients.locationsStorage(),
+      clients.institutionsStorage(),
+      clients.campusesStorage(),
+      clients.librariesStorage()
+    );
+  }
+
+  public CompletableFuture<Result<Location>> getLocation(Item item) {
     return SingleRecordFetcher.json(locationsStorageClient, "locations",
       response -> succeeded(null))
-      .fetch(item.getLocationId());
+      .fetch(item.getLocationId())
+      .thenApply(r -> r.map(Location::from))
+      .thenCompose(r -> r.after(this::loadLibrary))
+      .thenCompose(r -> r.after(this::loadCampus))
+      .thenCompose(r -> r.after(this::loadInstitution));
   }
 
-  public CompletableFuture<Result<Map<String, JsonObject>>> getLocations(
+  public CompletableFuture<Result<Map<String, Location>>> getLocations(
```

Comment:
Similarly for this method, please add the relevant permissions for each endpoint that uses this